model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/realm/entity/JpaRealmEntity.java

```diff
@@ -0,0 +1,1023 @@
+
+    @Id
+    @Column
+    private UUID id;
+
+    //used for implicit optimistic locking
+    @Version
+    @Column
+    private int version;
+
+    @Type(type = "jsonb")
+    @Column(columnDefinition = "jsonb")
+    private final JpaRealmMetadata metadata;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private Integer entityVersion;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private String name;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private String displayName;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private String displayNameHtml;
+
+    @Column(insertable = false, updatable = false)
+    @Basic(fetch = FetchType.LAZY)
+    private Boolean enabled;
+
+    @OneToMany(mappedBy = "root", cascade = CascadeType.PERSIST, orphanRemoval = true)
+    private final Set<JpaRealmAttributeEntity> attributes = new HashSet<>();
+
+    @OneToMany(mappedBy = "root", cascade = CascadeType.PERSIST, orphanRemoval = true)
+    private final Set<JpaComponentEntity> components = new HashSet<>();
+
+    /**
+     * No-argument constructor, used by hibernate to instantiate entities.
+     */
+    public JpaRealmEntity() {
+        this.metadata = new JpaRealmMetadata();
+    }
+
+    public JpaRealmEntity(final DeepCloner cloner) {
+        this.metadata = new JpaRealmMetadata(cloner);
+    }
+
+    /**
+     * Used by hibernate when calling cb.construct from read(QueryParameters) method.
+     * It is used to select realm without metadata(json) field.
+     */
+    public JpaRealmEntity(final UUID id, final int version, final Integer entityVersion, final String name,
+                          final String displayName, final String displayNameHtml, final Boolean enabled) {
+        this.id = id;
+        this.version = version;
+        this.entityVersion = entityVersion;
+        this.name = name;
+        this.displayName = displayName;
+        this.displayNameHtml = displayNameHtml;
+        this.enabled = enabled;
+        this.metadata = null;
+    }
+
+    public boolean isMetadataInitialized() {
+        return metadata != null;
+    }
+
+    @Override
+    public Integer getEntityVersion() {
+        if (isMetadataInitialized()) return metadata.getEntityVersion();
+        return entityVersion;
+    }
+
+    @Override
+    public void setEntityVersion(Integer entityVersion) {
+        metadata.setEntityVersion(entityVersion);
+    }
+
+    @Override
+    public Integer getCurrentSchemaVersion() {
+        return CURRENT_SCHEMA_VERSION_REALM;
+    }
+
+    @Override
+    public int getVersion() {
+        return version;
+    }
+
+    @Override
+    public String getId() {
+        return id == null ? null : id.toString();
+    }
+
+    @Override
+    public void setId(String id) {
+        String validatedId = UuidValidator.validateAndConvert(id);
```

Comment:
wondering if it should also be implemented for other areas as well ... if so perhaps we could move it into `JpaMapKeycloakTransaction.create` method, wdyt?

Reply:
create wouldn't work for entities such as the ComponentEntity because its creation happens when it is added to the components relationship - i.e. components are not created via `tx.create`. I would indeed put this check in every entity's `setId(String id)` because this is always called, whether from the `tx.create` or when creating an entity by adding it to a relationship (such as components or auth sessions that are added to the root auth session).

Reply:
Added the validation on `setId` to all other entities that have non-generated ids. For generated ids this method should never be called.